posts/2021/01/pytorch-tensoriterator-internals-update.md

```diff
@@ -0,0 +1,264 @@
+(which is an instance of the `loop_t` lambda) contains a `char*` pointer for
+each of the tensors, in the order that they are specified in the
+`TensorIteratorConfig`. To make the implementation agnostic of any particular
+data type, the pointer is typecast to `char`, so we can access it as an array
+of bytes.
+
+The second argument is `const int64_t* strides`, which is an array containing
+the strides of each tensor in the dimension that you're iterating over. We can
+add this stride to the pointer received in order to reach the next element in
+the tensor. The last argument is `int64_t n` which is the size of the dimension
+being iterated over.
+
+`for_each` implicitly parallelizes the operation by executing `copy_loop` in
+parallel if the number of iterations is more than the value of
+`internal::GRAIN_SIZE`, which is a value that is determined as the 'right
+amount' of data to iterate over in order to gain a significant speedup using
+multi-threaded execution. If you want to explicitly specify that your operation
+_must_ run in serial, then use the `serial_for_each` loop.
+
+``` cpp
+at::TensorIteratorConfig iter_config;
+iter_config
+  .add_output(out)
+  .add_input(a)
+
+  // call if output was already allocated
+  .resize_outputs(false)
+
+  // call if inputs/outputs have different types
+  .check_all_same_dtype(false);
+
+auto iter = iter_config.build();
+
+// Copies data from input into output
+auto copy_loop = [](char** data, const int64_t* strides, int64_t n) {
+  auto* out_data = data[0];
+  auto* in_data = data[1];
+
+  for (int i = 0; i < n; i++) {
+    // assume float data type for this example
+    *reinterpret_cast<float*>(out_data) += *reinterpret_cast<float*>(in_data);
+    out_data += strides[0];
+    in_data += strides[1];
+  }
+};
+
+iter.for_each(copy_loop);
+```
+
+### Using kernels for iterations
+
+Frequently we want to create a kernel that applies a simple point-wise function
+onto entire tensors. `TensorIterator` provides various such generic kernels
+that can be used for iterating over the elements of a tensor without having to
+worry about the stride, data type of the operands or details of the
+parallelism.
+
+For example, say we want to build a function that performs the point-wise
+addition of two tensors and stores the result in a third tensor. We can use the
+`cpu_kernel` function. Note that in this example we assume a tensor of `float`,
+but you can use one of the `AT_DISPATCH_ALL_TYPES*` macros to support multiple
+data types.
+
+``` cpp
+at::TensorIteratorConfig iter_config;
+iter_config
+  .add_output(c)
+  .add_input(a)
+  .add_input(b);
+
+auto iter = iter_config.build();
+
+// Element-wise add
+at::native::cpu_kernel(iter, [] (float a, float b) -> float {
+  return a + b;
+});
+```
+
+Writing the kernel in this way ensures that the value returned by the lambda
+passed to `cpu_kernel` will populate the corresponding position in the target
+output tensor.
```

Comment:
Only if `input` strictly broadcasts over `output`. That is `input` broadcasts over `output` and `output.shape >= input.shape`.